### PR TITLE
[Autodiff] Derivative Registration for the Get and Set Accessors

### PR DESCRIPTION
**Pull request by ematejska**
This implements the support for`@derivative` derivative registration for the get and set accessors (https://bugs.swift.org/browse/TF-1193.). The class set accessor is not yet supported.

Done together with @dan-zheng

### CODE REVIEW

**Comment by ematejska** [comment]
@swift-ci please smoke test

**Comment by ematejska** [comment]
@swift-ci please test

File: utils/gyb_syntax_support/AttributeNodes.py [hunk 1]
```diff
@@ -292,6 +292,15 @@
                    '''),
              Child('OriginalDeclName', kind='QualifiedDeclName',
                    description='The referenced original declaration name.'),
+             Child('Period', kind='PeriodToken',
+                   description='''
+                   The period separating the computed var or subscript
+                   and accessor name.
```
Comment (dan-zheng) [inline utils/gyb_syntax_support/AttributeNodes.py:298]: Minor changes for consistent verbiage:
```suggestion
                   The period separating the original declaration name and the
                   accessor name.
```

File: lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -1079,6 +1105,7 @@ static bool parseBaseTypeForQualifiedDeclName(Parser &P, TypeRepr *&baseType) {
 static bool parseQualifiedDeclName(Parser &P, Diag<> nameParseError,
                                    TypeRepr *&baseType,
                                    DeclNameRefWithLoc &original) {
+  {
```
Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp:1108]: Could you please indent the code that's within curly braces, to make the nesting clearer?
I recommend using [`git clang-format`](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) to automatically format patches.

File: lib/Parse/ParseDecl.cpp [hunk 3]
```diff
@@ -1092,7 +1119,23 @@ static bool parseQualifiedDeclName(Parser &P, Diag<> nameParseError,
       Parser::DeclNameFlag::AllowOperators);
   // The base type is optional, but the final unqualified declaration name is
   // not. If name could not be parsed, return true for error.
-  return !original.Name;
+  if (!original.Name)
+    return true;
+  }
+
+  // Parse to see if this is an accessor and set it's type.  This is an optional field.
```
Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp:1126]:
```suggestion
  // Parse an optional accessor kind.
```

Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp:1126]: If this isn't tested already, could you please add a `@derivative` attribute test case referencing the accessor of a non-storage declaration? This should be diagnosed.
```swift
func function(_ x: Float) -> Float {
  x
}

// Test invalid reference to an accessor of a non-storage declaration.
@derivative(of: function(_:).get)
func vjpFunction(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
  fatalError()
}
```

File: lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -1057,6 +1071,18 @@ static bool parseBaseTypeForQualifiedDeclName(Parser &P, TypeRepr *&baseType) {
...
+
+  // Check if this is a reference to an accessor in a computed property.
+  // FIXME: There is an ambiguity here because instead of a computed
+  // property with an accessor, this could be a type with a function
+  // name like an accessor.
+  if (P.Tok.is(tok::period)) {
+     const Token &nextToken = P.peekToken();
+     if (isAccessorLabel(nextToken) != None)
+       return false;
+  }
```
Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp:1083]: Could you please add a `@derivative` attribute test case exercising this ambiguity?
```swift
extension Float {
  // Original function name conflicts with an accessor name ("set").
  func set() -> Float {
    self
  }

  // Original function name does not conflict with an accessor name.
  func method() -> Float {
    self
  }

  // Test ambiguous parse.
  // Expected:
  // - Base type: `Float`
  // - Declaration name: `set`
  // - Accessor kind: <none>
  // Actual:
  // - Base type: <none>
  // - Declaration name: `Float`
  // - Accessor kind: `set`
  @derivative(of: Float.set)
  func jvpSet() -> (value: Float, differential: (Float) -> Float) {
    fatalError()
  }

  @derivative(of: Float.method)
  func jvpMethod() -> (value: Float, differential: (Float) -> Float) {
    fatalError()
  }
}
```

File: lib/Sema/TypeCheckAttr.cpp [hunk 5]
```diff
@@ -3609,12 +3610,14 @@ static IndexSubset *computeDifferentiabilityParameters(
 // If the function declaration cannot be resolved, emits a diagnostic and
 // returns nullptr.
 static AbstractFunctionDecl *findAbstractFunctionDecl(
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp:3612]: `findAbstractFunctionDecl` is in need of a revamp, diagnostics are currently handled messily. I'll look into it after this PR lands.

**Review by dan-zheng — COMMENTED** [review]
Nice work!

**Review by dan-zheng — COMMENTED** [review]
When syntax definitions are changed, I believe `apple/swift-syntax` needs to be updated with newly generated syntax files. This is necessary to pass [`--swiftsyntax-verify-generated-files`](https://github.com/apple/swift/blob/7802d1755cea1e5e6207e12f92bf036d7c490b73/utils/build_swift/build_swift/driver_arguments.py#L570).

[Here's an example PR](https://github.com/apple/swift-syntax/pull/188). I think you can regenerate `swift-syntax` files locally with `utils/build-script --infer --swiftsyntax`.

File: lib/Sema/TypeCheckAttr.cpp [hunk 6]
```diff
@@ -4429,6 +4446,13 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
                    diag::autodiff_attr_original_decl_invalid_kind,
                    originalName.Name);
   };
+  auto missingAccessorDiagnostic = [&]() {
+    auto accessorKind = originalName.AccessorKind.getValueOr(AccessorKind::Get);
+    auto accessorLabel = getAccessorLabel(accessorKind);
+    diags.diagnose(originalName.Loc, diag::autodiff_attr_accessor_not_found,
+             originalName.Name, accessorLabel);
```
Comment (rxwei) [inline lib/Sema/TypeCheckAttr.cpp:4453]: Nit: Fix indentation.

File: test/AutoDiff/Sema/derivative_attr_type_checking.swift [hunk 7]
```diff
@@ -442,24 +463,129 @@ extension Struct {
...
+  // This requires changing derivative type calculation rules for functions with
+  // class-typed parameters. We need to assume that all functions taking
+  // class-typed operands may mutate those operands.
+  // expected-error @+1 {{cannot yet register derivative for class property or subscript setters}}
+  @derivative(of: subscript.set)
+  func vjpSubscriptSetter(_ newValue: Float) -> (
+    value: (), pullback: (inout TangentVector) -> Float
+  ) {
+    fatalError()
+  }
```
Comment (dan-zheng) [inline test/AutoDiff/Sema/derivative_attr_type_checking.swift:588]: In addition to `@derivative` attribute type-checking tests, we should add some end-to-end differentiation tests to verify that accessor derivative registration fully works.

Comment (dan-zheng) [reply test/AutoDiff/Sema/derivative_attr_type_checking.swift:588]: Done in f1f60345b47d729d34afcd3ba52f20cb8b149f68.

File: lib/Parse/ParseDecl.cpp (hunk 2 above)
Comment (ematejska) [reply lib/Parse/ParseDecl.cpp:1108]: Dan did this himself.  Thanks!

File: lib/Parse/ParseDecl.cpp (hunk 4 above)
Comment (dan-zheng) [reply lib/Parse/ParseDecl.cpp]: I don't believe `@derivative(of: Swift.Float.set)` parses correctly today due to excessive backtracking in `parseBaseTypeForQualifiedDeclName`. I clarified the FIXME comment in https://github.com/apple/swift/pull/32614/commits/bf998e1559a36dfe938a17c33f77b148a60216f5 to explain why this occurs.

I think it's fine to land this initial patch without supporting this edge case, but we should add it as a test!

**Comment by gottesmm** [comment]
Before this lands, can you be sure to re-base and clean up the git history by smooshing commits that don't stand on their own. Thanks!

File: lib/Parse/ParseDecl.cpp (hunk 4 above)
Comment (ematejska) [reply lib/Parse/ParseDecl.cpp]: Added

File: lib/Parse/ParseDecl.cpp (hunk 3 above)
Comment (ematejska) [reply lib/Parse/ParseDecl.cpp]: Added.  Good one.

**Comment by ematejska** [comment]
@swift-ci please smoke test

**Comment by ematejska** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckAttr.cpp [hunk 8]
```diff
@@ -3668,6 +3668,8 @@ static AbstractFunctionDecl *findAbstractFunctionDecl(
           missingAccessor = true;
       } else
         candidate = asd->getAccessor(AccessorKind::Get);
+    } else if (accessorKind != None) {
+        missingAccessor = true;
```
Comment (dan-zheng) [inline lib/Sema/TypeCheckAttr.cpp:3668]: minor, we can fix this later:
```suggestion
      missingAccessor = true;
```

**Review by dan-zheng — APPROVED** [review]
Nice!

@ematejska: would you like to preemptively squash commits? Or you could "Squash and merge" with a short commit description.

**Review by rxwei — APPROVED** [review]

**Comment by ematejska** [comment]
@swift-ci please smoke test

**Comment by ematejska** [comment]
@swift-ci please test

**Comment by ematejska** [comment]
@swift-ci please test os x platform

**Comment by ematejska** [comment]
@swift-ci please smoke test os x platform

**Comment by ematejska** [comment]
> Nice!
> 
> @ematejska: would you like to preemptively squash commits? Or you could "Squash and merge" with a short commit description.

I'll do the Squash and merge.

**Comment by ematejska** [comment]
> When syntax definitions are changed, I believe `apple/swift-syntax` needs to be updated with newly generated syntax files. This is necessary to pass [`--swiftsyntax-verify-generated-files`](https://github.com/apple/swift/blob/7802d1755cea1e5e6207e12f92bf036d7c490b73/utils/build_swift/build_swift/driver_arguments.py#L570).
> 
> [Here's an example PR](https://github.com/apple/swift-syntax/pull/188). I think you can regenerate `swift-syntax` files locally with `utils/build-script --infer --swiftsyntax`.

I'll try to do this today.

**Comment by ematejska** [comment]
Please test with:
https://github.com/apple/swift-syntax/pull/225

@swift-ci please test os x platform

**Pull request merged**